File: 0029-divide-two-integers/0029-divide-two-integers.java
class Solution {
    public int divide(int dividend, int divisor) {
        boolean sign = true;
        if((dividend>=0 && divisor<0)||(dividend<=0 && divisor>0)){
            sign = false;
        }

        long a = Math.abs((long)dividend);
        long b = Math.abs((long)divisor);
        long quotient = 0;

        while(a>=b){
            int count = 0;
            while(a>=b<<(count+1)){
                count++;
            }
            quotient += 1L << count;
            a -= b <<count;
        }

        if(quotient>=1L<<31){
            return sign ? Integer.MAX_VALUE: Integer.MIN_VALUE;
        }

        return sign ? (int)quotient: (int)-quotient;
    }
}

File: 0029-divide-two-integers/0029-divide-two-integers.py
class Solution(object):
    def divide(self, dividend, divisor):
        if dividend == divisor:
            return 1
        
        sign = True

        if (dividend >=0 and divisor<0) or (dividend<=0 and divisor>0):
            sign = False

        a= abs(dividend)
        b=abs(divisor)
        quotient = 0

        while a>=b:
            count =0
            while a>=b<<(count+1):
                count+=1
            quotient += 1<<count
            a -= b<<count

        if quotient>=(1<<31):
            return (2**31)-1 if sign else -(2**31)
        
        return quotient if sign else -quotient

        